Start to keep Architectural Decision Records (ADRs) for tuf reference implementation

joshuagl opened the pull request:
Fixes #1141, #1125 

**Description of the changes being introduced by the pull request**:

Add initial Architectural Decision Records (ADRs) documenting some workflow decisions before the [refactor](https://github.com/theupdateframework/tuf/projects/2):
1. Use of MADR format to keep ADRs for the reference implementation
2. Decision to adopt Python 3.6+ moving forward (dropping Python 2.7)

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

joshuagl committed:
docs/adr: start to keep ADRs in MADR format

In order to make decisions about the code and the design explicit and easier
to reference in future we want to record significant architectural decisions.

This commit introduces docs/adr with a template Architectural Decision Record
and index using the [MADR](https://adr.github.io/madr/) format.

It also adds ADR 0000 to document the decisions to use MADR.

Fixes #1141

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
ADR: only use Python 3.6+

Document the decision drop support for EOL Python versions, most notable
Python 2.7

Fixes #1125

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Teach git to ignore Emacs backup files

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl commented:
I'm not sure why the AppVeyor build is failing to find dateutil when other recent builds have successfully run the same tests which are now failing, i.e. AppVeyor run for the 0.15.0 release PR #1183 is [green](https://ci.appveyor.com/project/lukpueh/tuf-fdnro/builds/35873166) while this one is [red](https://ci.appveyor.com/project/lukpueh/tuf-fdnro/builds/35875709)

jku commented:
> I'm not sure why the AppVeyor build is failing to find dateutil when other recent builds have successfully run the same tests which are now failing, i.e. AppVeyor run for the 0.15.0 release PR #1183 is [green](https://ci.appveyor.com/project/lukpueh/tuf-fdnro/builds/35873166) while this one is [red](https://ci.appveyor.com/project/lukpueh/tuf-fdnro/builds/35875709)

The securesystemslib release probably has something to do with it?

joshuagl commented:
> > I'm not sure why the AppVeyor build is failing to find dateutil when other recent builds have successfully run the same tests which are now failing, i.e. AppVeyor run for the 0.15.0 release PR #1183 is [green](https://ci.appveyor.com/project/lukpueh/tuf-fdnro/builds/35873166) while this one is [red](https://ci.appveyor.com/project/lukpueh/tuf-fdnro/builds/35875709)
> 
> The securesystemslib release probably has something to do with it?

I think you're right! We did remove `python-dateutil` as a dependency of securesystemslib.

mnm678 reviewed (approved):
These look great. Once the AppVeyor issue is resolved I think it's good to go.

joshuagl committed:
ADR0001: clarify when/where Python 3.6+ is expected

Provide additional context to clarify where we expect Python 3.6+ to be used
exclusively (new modules) and link to other discussions around the future of
Python 2.7 supporting code.

Signed-off-by: Joshua Lock <jlock@vmware.com>

lukpueh reviewed (approved):
Mille grazie, @joshuagl!
